src/query/storages/fuse/src/operations/mutation/mutation_source.rs

```diff
@@ -351,40 +352,61 @@ impl Processor for MutationSource {
         match std::mem::replace(&mut self.state, State::Finish) {
             State::ReadData(Some(part)) => {
                 let settings = ReadSettings::from_ctx(&self.ctx)?;
-                let part = MutationPartInfo::from_part(&part)?;
-
-                self.index = BlockMetaIndex {
-                    segment_idx: part.index.segment_idx,
-                    block_idx: part.index.block_idx,
-                };
-                if matches!(self.action, MutationAction::Deletion) {
-                    self.stats_type = ClusterStatsGenType::WithOrigin(part.cluster_stats.clone());
-                }
+                let res = MutationPartInfo::from_part(&part);
+                if let Ok(part) = res {
+                    self.index = BlockMetaIndex {
+                        segment_idx: part.index.segment_idx,
+                        block_idx: part.index.block_idx,
+                    };
+                    if matches!(self.action, MutationAction::Deletion) {
+                        self.stats_type =
+                            ClusterStatsGenType::WithOrigin(part.cluster_stats.clone());
+                    }
 
-                let inner_part = part.inner_part.clone();
-                let fuse_part = FusePartInfo::from_part(&inner_part)?;
+                    let inner_part = part.inner_part.clone();
+                    let fuse_part = FusePartInfo::from_part(&inner_part)?;
 
-                if part.whole_block_mutation && matches!(self.action, MutationAction::Deletion) {
-                    // whole block deletion.
+                    if part.whole_block_mutation && matches!(self.action, MutationAction::Deletion)
+                    {
+                        // whole block deletion.
+                        let progress_values = ProgressValues {
+                            rows: fuse_part.nums_rows,
+                            bytes: 0,
+                        };
+                        self.ctx.get_write_progress().incr(&progress_values);
+                        let meta =
+                            SerializeDataMeta::create(self.index.clone(), self.stats_type.clone());
+                        self.state = State::Output(
+                            self.ctx.get_partition(),
+                            DataBlock::empty_with_meta(meta),
+                        );
+                    } else {
+                        let read_res = self
+                            .block_reader
+                            .read_columns_data_by_merge_io(
+                                &settings,
+                                &fuse_part.location,
+                                &fuse_part.columns_meta,
+                            )
+                            .await?;
+                        self.state = State::FilterData(inner_part, read_res);
+                    }
+                } else {
```

Comment:
is this the branch cover the case that  `MutationPartInfo::from_part(&part).is_error()`?

it works but seems to be not that "idiomatic",  maybe `MutationPartInfo` should be refactored into an Enum?